radix-engine/src/kernel/call_frame.rs

```diff
@@ -944,86 +950,79 @@ impl<L: Clone> CallFrame<L> {
+            return Err(CallFrameScanSubstatesError::NodeNotVisible(node_id.clone()));
         }
 
-        let (substates, store_access) = if heap.contains_node(node_id) {
+        let (keys, store_access) = if heap.contains_node(node_id) {
             (
-                heap.scan_substates(node_id, partition_num, count),
+                heap.scan_keys(node_id, partition_num, count),
                 StoreAccessInfo::new(),
             )
         } else {
-            store.scan_substates(node_id, partition_num, count)
+            store.scan_keys::<K>(node_id, partition_num, count)
         };
 
-        for substate in &substates {
-            for reference in substate.references() {
-                if reference.is_global() {
-                    self.stable_references
-                        .insert(reference.clone(), StableReferenceType::Global);
-                } else {
-                    // FIXME: check if non-global reference is needed
-                }
-            }
-        }
-
-        Ok((substates, store_access))
+        Ok((keys, store_access))
     }
 
-    pub fn take_substates<'f, S: SubstateStore>(
+    pub fn drain_substates<'f, K: SubstateKeyContent, S: SubstateStore>(
         &mut self,
         node_id: &NodeId,
         partition_num: PartitionNumber,
         count: u32,
         heap: &'f mut Heap,
         store: &'f mut S,
-    ) -> Result<(Vec<IndexedScryptoValue>, StoreAccessInfo), CallFrameTakeSortedSubstatesError>
-    {
+    ) -> Result<
+        (Vec<(SubstateKey, IndexedScryptoValue)>, StoreAccessInfo),
+        CallFrameDrainSubstatesError,
+    > {
         // Check node visibility
         if !self.get_node_visibility(node_id).can_be_read_or_write() {
-            return Err(CallFrameTakeSortedSubstatesError::NodeNotVisible(
+            return Err(CallFrameDrainSubstatesError::NodeNotVisible(
                 node_id.clone(),
             ));
         }
 
         let (substates, store_access) = if heap.contains_node(node_id) {
             (
-                heap.take_substates(node_id, partition_num, count),
+                heap.drain_substates(node_id, partition_num, count),
                 StoreAccessInfo::new(),
             )
         } else {
-            store.take_substates(node_id, partition_num, count)
+            store.drain_substates::<K>(node_id, partition_num, count)
         };
 
-        for substate in &substates {
+        for (_key, substate) in &substates {
             for reference in substate.references() {
                 if reference.is_global() {
                     self.stable_references
                         .insert(reference.clone(), StableReferenceType::Global);
                 } else {
-                    // FIXME: check if non-global reference is needed
+                    return Err(CallFrameDrainSubstatesError::CantDropOwnedNode(
+                        reference.clone(),
```

Comment:
Wrong error message here? It's about the substate containing non-global reference.